epoch.go

```diff
@@ -255,13 +258,25 @@ func newEpoch(newEpochConfig *pb.EpochConfig, checkpointTracker *checkpointTrack
 		proposer:          proposer,
 		sequences:         sequences,
 		lowestUnallocated: make([]int, len(config.buckets)),
+		logger:            myConfig.Logger,
+	}
+}
+
+func (e *epoch) getSequence(seqNo uint64) (*sequence, error) {
+	if seqNo < e.lowWatermark() || seqNo > e.highWatermark() {
+		return nil, errors.Errorf("requested seq no (%d) is out of range [%d - %d]",
+			seqNo, e.lowWatermark(), e.highWatermark())
 	}
+	return e.sequences[int(seqNo-e.lowWatermark())], nil
 }
 
 func (e *epoch) applyPreprepareMsg(source NodeID, seqNo uint64, batch []*request) *Actions {
-	baseCheckpoint := e.baseCheckpoint
-	offset := int(seqNo-baseCheckpoint.SeqNo) - 1
-	seq := e.sequences[offset]
+	seq, err := e.getSequence(seqNo)
+	if err != nil {
+		e.logger.Error(err.Error())
+		return nil
```

Comment:
You'll notice in these methods, we never want to actually return nil, we always return an empty Actions instead.  This is so the caller can blindly accumulate the actions without worrying about nil pointer dereferences.  This path would trigger a nil dereference when the sequence is out of range (same goes for the rest).